File: platforms/android/ime/src/main/java/app/funput/funput/ime/editing/gestures/SpaceTapTracker.kt
package app.funput.funput.ime.editing.gestures

/**
 * Recognizes two space presses in quick succession.
 *
 * A value type over an injected clock, so the double-tap window is testable without
 * waiting on wall time.
 */
internal class SpaceTapTracker(
    private val doubleTapIntervalMillis: Long = DefaultIntervalMillis,
    private val clock: () -> Long = { System.nanoTime() / 1_000_000L },
) {
    private var lastTapTime: Long? = null

    init {
        require(doubleTapIntervalMillis > 0L) { "Double-tap interval must be positive" }
    }

    /**
     * Records a space press and reports whether it completes a double tap.
     *
     * A firing press clears the sequence, so three spaces produce one substitution
     * and a plain space rather than a second one.
     */
    fun registerSpace(): Boolean {
        val tapTime = clock()
        val last = lastTapTime
        if (last != null && tapTime - last <= doubleTapIntervalMillis) {
            reset()
            return true
        }
        lastTapTime = tapTime
        return false
    }

    /**
     * Called for every non-space key: two spaces separated by other input are not a
     * double tap, however fast they arrive.
     */
    fun reset() {
        lastTapTime = null
    }

    private companion object {
        const val DefaultIntervalMillis = 300L
    }
}

File: platforms/ios/Packages/FunputKit/Sources/KeyboardInput/State/Gestures/SpaceTapTracker.swift
import Foundation

/// Recognizes two space presses in quick succession.
///
/// Deliberately shaped like ``ShiftStateController``: a value type over an injected
/// clock, so the double-tap window is testable without waiting on wall time.
@MainActor
struct SpaceTapTracker {
    private let doubleTapInterval: TimeInterval
    private let clock: () -> TimeInterval
    private var lastTapTime: TimeInterval?

    init(
        doubleTapInterval: TimeInterval = 0.3,
        clock: @escaping () -> TimeInterval = { ProcessInfo.processInfo.systemUptime }
    ) {
        self.doubleTapInterval = doubleTapInterval
        self.clock = clock
    }

    /// Records a space press and reports whether it completes a double tap.
    ///
    /// A firing press clears the sequence, so three spaces produce one substitution
    /// and a plain space rather than a second one.
    mutating func registerSpace() -> Bool {
        let tapTime = clock()
        if let lastTapTime, tapTime - lastTapTime <= doubleTapInterval {
            reset()
            return true
        }
        lastTapTime = tapTime
        return false
    }

    /// Called for every non-space key: two spaces separated by other input are not a
    /// double tap, however fast they arrive.
    mutating func reset() {
        lastTapTime = nil
    }
}
